langfuse/decorators/langfuse.py

```diff
@@ -0,0 +1,388 @@
+from langfuse.llama_index import LlamaIndexCallbackHandler
+from langfuse.types import ObservationParams, SpanLevel
+from langfuse.utils import _get_timestamp
+
+
+langfuse_context: ContextVar[Optional[Langfuse]] = ContextVar(
+    "langfuse_context", default=None
+)
+observation_stack_context: ContextVar[
+    List[Union[StatefulTraceClient, StatefulSpanClient]]
+] = ContextVar("observation_stack_context", default=[])
+observation_params_context: ContextVar[
+    DefaultDict[str, ObservationParams]
+] = ContextVar(
+    "observation_params_context",
+    default=defaultdict(
+        lambda: {
+            "name": None,
+            "user_id": None,
+            "session_id": None,
+            "version": None,
+            "release": None,
+            "metadata": None,
+            "tags": None,
+            "input": None,
+            "output": None,
+            "level": None,
+            "status_message": None,
+            "start_time": None,
+            "end_time": None,
+        },
+    ),
+)
+
+
+class LangfuseDecorator:
+    log = logging.getLogger("langfuse")
+
+    def __init__(self):
+        self._langfuse: Optional[Langfuse] = None
+
+    def trace(self, func: Callable) -> Callable:
+        """
+        Wraps a function to automatically create and manage Langfuse tracing around its execution.
+
+        This decorator captures the start and end times of the function, along with input parameters and output results, and automatically updates the observation context.
+        It creates traces for top-level function calls and spans for nested function calls, ensuring that each observation is correctly associated with its parent observation.
+        In the event of an exception, the observation is updated with error details.
+
+        Usage:
+            @langfuse.trace\n
+            def your_function_name(args):
+                # Your function implementation
+
+        Parameters:
+            func (Callable): The function to be wrapped by the decorator.
+
+        Returns:
+            Callable: A wrapper function that, when called, executes the original function within a Langfuse observation context.
+
+        Raises:
+            Exception: Propagates any exceptions raised by the wrapped function, after logging and updating the observation with error details.
+
+        Note:
+            - The decorator automatically manages observation IDs and context, but you can manually specify an observation ID using the `langfuse_observation_id` keyword argument when calling the wrapped function.
+            - To update observation or trace parameters (e.g., metadata, session_id), use `langfuse.update_current_observation` and `langfuse.set_current_trace_params` within the wrapped function.
+        """
+
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            langfuse = self._get_langfuse()
+            stack = observation_stack_context.get().copy()
+            parent = stack[-1] if stack else None
```

Comment:
Quick question: say i have a function and annotate it with a trace. Within that function i start two threads, which execute two annotated functions. Will both get the same parent?